src/client/components/pinboard-tile/pinboard-tile.tsx

```diff
@@ -0,0 +1,255 @@
+import { SortOn } from "../../../common/models/sort-on/sort-on";
+import { Timekeeper } from "../../../common/models/timekeeper/timekeeper";
+import { DatasetLoad, error, isError, isLoaded, isLoading, loaded, loading } from "../../../common/models/visualization-props/visualization-props";
+import { debounceWithPromise, Unary } from "../../../common/utils/functional/functional";
+import { Fn } from "../../../common/utils/general/general";
+import { MAX_SEARCH_LENGTH } from "../../config/constants";
+import { setDragData, setDragGhost } from "../../utils/dom/dom";
+import { DragManager } from "../../utils/drag-manager/drag-manager";
+import { reportError } from "../../utils/error-reporter/error-reporter";
+import { Loader } from "../loader/loader";
+import { QueryError } from "../query-error/query-error";
+import { SearchableTile } from "../searchable-tile/searchable-tile";
+import { PinboardDataset } from "./pinboard-dataset";
+import { pinboardIcons } from "./pinboard-icons";
+import "./pinboard-tile.scss";
+import { EditMode, RowsMode } from "./utils/edit-mode";
+import { isClauseEditable } from "./utils/is-clause-editable";
+import { isDimensionPinnable } from "./utils/is-dimension-pinnable";
+import { makeQuery } from "./utils/make-query";
+import { isPinnableClause, PinnableClause } from "./utils/pinnable-clause";
+import { equalParams, QueryParams } from "./utils/query-params";
+import { shouldFetchData } from "./utils/should-fetch";
+import { tileStyles } from "./utils/tile-styles";
+
+export class PinboardTileProps {
+  essence: Essence;
+  clicker: Clicker;
+  dimension: Dimension;
+  timekeeper: Timekeeper;
+  sortOn: SortOn;
+  onClose: Fn;
+}
+
+export interface PinboardTileState {
+  searchText: string;
+  showSearch: boolean;
+  datasetLoad: DatasetLoad;
+}
+
+const noMeasureError = new Error("No measure selected");
+
+export class PinboardTile extends React.Component<PinboardTileProps, PinboardTileState> {
+
+  state: PinboardTileState = {
+    searchText: "",
+    showSearch: false,
+    datasetLoad: this.props.sortOn ? loading : error(noMeasureError)
+  };
+
+  private loadData(params: QueryParams) {
+    if (!params.sortOn) {
+      this.setState({ datasetLoad: error(noMeasureError) });
+      return;
+    }
+    this.setState({ datasetLoad: loading });
+    this.fetchData(params)
+      .then(loadedDataset => {
+        // TODO: encode it better
+        // null is here when we get out of order request, so we just ignore it
+        if (!loadedDataset) return;
+        this.setState({ datasetLoad: loadedDataset });
+      });
+  }
+
+  private fetchData(params: QueryParams): Promise<DatasetLoad | null> {
+    this.lastQueryParams = params;
+    return this.debouncedCallExecutor(params);
+  }
+
+  private lastQueryParams: Partial<QueryParams> = {};
+
+  private callExecutor = (params: QueryParams): Promise<DatasetLoad | null> => {
```

Comment:
Why this particular component has to be responsible for error handling? It looks very generic.

Reply:
I agree that it looks like something to be extracted, but I don't see how to do it well. I prefer copy-paste here instead of some over engineered abstraction. 

I don't see what should handle error handling instead, though.